### ISSUE TITLE
incorrect length in JSON?

### ISSUE THREAD

**Issue by jjn1056** [perl-catalyst/catalyst-runtime#143]
Hi!
if content type is 'application/json' or 'application/json; charset=utf-8' Catalyst sets content length in chars, NOT IN BYTES and I'm getting 
{"id":1, "msg":"В Питере
if content type is 'text/html' Catalyst sets content length in bytes (properly) and everything works fine
Is there any workaround to configure this behaviour, except setting content length manually everytime ?

my $json_text = '{"id":1, "msg":"В Питере пить"}';
$c->response->content_type('application/json');
$c->response->content_length(bytes::length $json_text);
$c->response->body($json_text);
Thanks in advance

**Comment by jjn1056** [comment]
Looks like to a bug to me, although I'm not personally keen on the auto length setting in Catalyst it should be corrected.  I'm happy to get a patch, or at the very least give me a broken test case (checkout https://github.com/perl-catalyst/catalyst-runtime/blob/master/t/utf_incoming.t 

and see if you can help me figure it out -jnap

**Comment by melmothx** [comment]
If you ask me, this is not a bug. application/json is not encoded by catalyst (as per doc) so it must be encoded to bytes before setting the body. Most serializers do that, so the wrong behavior here is just because of the manual json creation.

**Comment by jjn1056** [comment]
@melmothx good catch, you'd think I'd remember that since I wrote the doc and made that call for back compat reasons.  I wonder if the docs could not be more obvious and maybe I should add a config setting to 'force_json_encode_always' for people that are rolling their own... I guess I was thinking if you are doing that yourself I assume you know what you are doing...

**Comment by jjn1056** [comment]
I'm thinking the docs should at least recommend one considers setting content length in your custom view, just to be safe.

**Comment by melmothx** [comment]
John Napiorkowski <notifications@github.com> writes:

> I'm thinking the docs should at least recommend one considers setting
> content length in your custom view, just to be safe.

I believe that if the string is passed correctly, the content length is
set by catalyst correct as well (or at least works for me so). In the OP
report, the length was off because chars were passed and setting the
length was not going to make the situation better (because of broken
encoding).

-- 
Marco
